File: MonitoreoCNET/MonitoreoCNET/Models/Persona/Miembro.cs

using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using Dominio.Entidades.Ubicacion;
using MonitoreoCNET.Models.Comunidades;

namespace MonitoreoCNET.Models.Persona
{
    public enum TipoMiembro
    {
        OBSERVADOR,
        AFECTADO
    }

    public enum MedioDeNotificacion
    {
        MAIL,
        WHATSAPP
    }

    public enum EstrategiaDeNotificacion
    {
        CUANDOSUCEDEN,
        SINAPUROS
    }

[Table("miembro")]
    public class Miembro
    {
        [Key]
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public int Id { get; set; }

        [Column("nombre")]
        public string Nombre { get; set; }

        [Column("apellido")]
        public string Apellido { get; set; }

        //[OneToOne]
        public virtual Usuario Usuario { get; set; }

        [ForeignKey("id_localizacion")]
        public virtual Localizacion LocalizacionDeInteres { get; set; }

        //[OneToMany(nameof(Interes.Miembro))]
        public virtual List<Interes> Intereses { get; set; }

        [ForeignKey("id_comunidad")]
        public virtual Comunidad Comunidad { get; set; }

        [Column("medio_notificacion")]
        public MedioDeNotificacion MedioDeNotificacion { get; set; }

        [Column("estrategia_notificacion")]
        public EstrategiaDeNotificacion EstrategiaDeNotificacion { get; set; }

        [Column("telefono")]
        public string Telefono { get; set; }

        [Column("tipo_miembro")]
        public TipoMiembro Tipo { get; set; }

        public Miembro(string nombre, string apellido, Localizacion localizacion)
        {
            Nombre = nombre;
            Apellido = apellido;
            LocalizacionDeInteres = localizacion;
            Intereses = new List<Interes>();
        }

        public Miembro()
        {
            Intereses = new List<Interes>();
        }

        public void AgregarInteres(Interes interes)
        {
            Intereses.Add(interes);
            interes.Miembro = this;
        }

        public void RemoverInteres(Interes interes)
        {
            Intereses.Remove(interes);
        }
        /*
        public bool EstaCercaDeIncidente(Localizacion localizacion)
        {
            if ((localizacion.EsProvincia && LocalizacionDeInteres.EsProvincia) || (!localizacion.EsProvincia && !LocalizacionDeInteres.EsProvincia))
            {
                return LocalizacionDeInteres.Id == localizacion.Id;
            }
            else if (localizacion.EsProvincia)
            {
                return LocalizacionDeInteres.Provincia.Id == localizacion.Id;
            }
            else
            {
                return LocalizacionDeInteres.Id == localizacion.Provincia.Id;
            }
        }*/

        public void SetLocalizacionDeInteres(Localizacion localizacion/*, Notificador notificador*/)
        {
            LocalizacionDeInteres = localizacion;

            if (Comunidad == null)
                return;
            /*
            foreach (var incidente in Comunidad.IncidentesSegunEstado(Estado.ABIERTO))
            {
                if (EstaCercaDeIncidente(incidente.Localizacion))
                {
                    notificador.NotificarAMiembro(this, new RevisionIncidente(), incidente);
                }
            }*/
        }
    }
}

File: monitoreo/src/main/java/models/entidades/persona/Miembro.java
package models.entidades.persona;

import models.entidades.comunidad.Comunidad;
import models.entidades.incidente.Estado;
import models.entidades.incidente.Incidente;
import models.entidades.ubicacion.Localizacion;
import lombok.Getter;
import lombok.Setter;
import models.entidades.persona.autentificacion.Usuario;
import models.modulos.Notificador.Notificaciones.Notificacion.RevisionIncidente;
import models.modulos.Notificador.Notificador;

import javax.persistence.*;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name="miembro")
public class Miembro {

  @Id
  @GeneratedValue(strategy = GenerationType.IDENTITY)
  @Getter @Setter
  private int id;

  @Getter @Setter
  @Column(name = "nombre")
  private String nombre;

  @Getter @Setter
  @Column(name = "apellido")
  private String apellido;

  @Getter @Setter
  @Column(name = "dni")
  private String dni;

  @Getter @Setter
  @OneToOne(cascade = {CascadeType.PERSIST, CascadeType.MERGE, CascadeType.REMOVE}, fetch = FetchType.LAZY)
  private Usuario usuario;

  @Getter @Setter
  @ManyToOne(fetch = FetchType.LAZY)
  @JoinColumn(name = "id_localizacion", referencedColumnName = "id")
  private Localizacion localizacionDeInteres;

  @Getter @Setter
  @OneToMany(mappedBy = "miembro", cascade = {CascadeType.PERSIST, CascadeType.MERGE, CascadeType.REMOVE}, fetch = FetchType.LAZY)
  private List<Interes> intereses;

  @Getter @Setter
  @ManyToOne(fetch = FetchType.LAZY)
  @JoinColumn(name="id_comunidad",referencedColumnName = "id")
  private Comunidad comunidad;

  @Getter @Setter
  @Column(name = "medio_notificacion")
  @Enumerated(EnumType.STRING)
  private MedioDeNotificacion medioDeNotificacion;

  @Getter @Setter
  @Column(name = "estrategia_notificacion")
  @Enumerated(EnumType.STRING)
  private EstrategiaDeNotificacion estrategiaDeNotificacion;

  @Getter @Setter
  @Column(name = "telefono")
  private String telefono;

  @Getter @Setter
  @Column(name = "tipo_miembro")
  @Enumerated(EnumType.STRING)
  private TipoMiembro tipo;

  public Miembro(String nombre, String apellido, Localizacion localizacion) {
    this.setNombre(nombre);
    this.setApellido(apellido);
    this.localizacionDeInteres = localizacion;
    this.intereses = new ArrayList<>();
    this.tipo = TipoMiembro.AFECTADO;
  }

  public Miembro(){
    this.intereses = new ArrayList<>();
    this.tipo = TipoMiembro.AFECTADO;
  }

  public void agregarInteres(Interes interes) {
    this.intereses.add(interes);
    interes.setMiembro(this);
  }

  public void removerInteres(Interes interes) {
    this.intereses.remove(interes);
  }

  public Boolean estaCercaDeIncidente(Localizacion localizacion) {

    if ((localizacion.getEsProvincia() && localizacionDeInteres.getEsProvincia()) || (!localizacion.getEsProvincia() && !localizacionDeInteres.getEsProvincia())) {
      return this.localizacionDeInteres.getId().equals(localizacion.getId());
    } else if (localizacion.getEsProvincia()) {
      return this.localizacionDeInteres.getProvincia().getId().equals(localizacion.getId());
    } else {
      return this.localizacionDeInteres.getId().equals(localizacion.getProvincia().getId());
    }

  }

  public void setLocalizacionDeInteres(Localizacion localizacion, Notificador notificador) {
    this.localizacionDeInteres = localizacion;

    if(this.comunidad == null)
        return;

    for (Incidente incidente : comunidad.incidentesSegunEstado(Estado.ABIERTO)) {
      if (this.estaCercaDeIncidente(incidente.getLocalizacion())) {
        notificador.notificarAMiembro(this, new RevisionIncidente(), incidente);
      }
    }
  }
}
